UI :  Added Bot Icon for bots across the UI.

NotVishesh opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #26831

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
For easy to differentiate between normal user and a bot . a bot icon is added at right of the bot
Continuing with the PR  #28561 and removed merge conflicts

**Screenshots and screen captures:**
<details>
<summary>Compose pills</summary>

![image](https://github.com/user-attachments/assets/d0ff00ba-cb23-43a7-9c25-fca1ae5b13f6)

</details>

<details>
<summary>Sidebar UI</summary>

![image](https://github.com/user-attachments/assets/e763d772-a322-4720-b511-03c69b85c96b)

</details>


<details>

<summary>Channels setting</summary>

![image](https://github.com/user-attachments/assets/57c66a24-4b23-4fd7-b3a6-628f423e6216)

</details>

<details>

<summary>DM headers</summary>

![image](https://github.com/user-attachments/assets/17a53dd8-7ca8-4f51-8020-7826d5cb9fba)

</details>

<details>

<summary>Bot Mention pills
</summary>

![image](https://github.com/user-attachments/assets/a138c0cc-3145-4539-9aed-b4d2337630d2)

</details>



<details>

<summary>Dark theme screenshots</summary>

![image](https://github.com/user-attachments/assets/bc28112c-aebe-418b-b11e-f668736dea3e)
<hr></hr>

![image](https://github.com/user-attachments/assets/bd9be181-811b-4c5e-a6e1-2c35c87bc247)
<hr></hr>

![image](https://github.com/user-attachments/assets/d57af7ce-bba7-4bc9-8099-2336807abba1)

<hr></hr>

![image](https://github.com/user-attachments/assets/4ba2c8bc-fd9a-4ced-99e4-4de91e8069d8)

<hr></hr>

![image](https://github.com/user-attachments/assets/c6d382e3-25b5-416c-989a-a08bc22db7ea)


</details>

<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! The screenshots mostly look reasonable, other than (1) the icon color in the selected line in the typeahead, and (2) the alignment in the bot mention pill (too high).

alya commented:
Please include light and dark theme screenshots for all the places where the icon was added, make sure the tests pass, and post a comment when ready for another round of review.

NotVishesh commented:
@alya Thanks for the feedback, I Just made those desired changes and also updated the screenshots.

timabbott commented on web/src/user_pill.ts:
```diff
@@ -84,6 +86,7 @@ export function create_item_from_email(
         deactivated: false,
         status_emoji_info,
         should_add_guest_user_indicator: people.should_add_guest_user_indicator(user.user_id),
+        user_is_bot: user_is_bot(user.user_id),
```
Can't one check `user.id_bot` directly? I don't see why you'd need a new function for this.

timabbott commented on web/templates/mention_content_wrapper.hbs:
```diff
@@ -1,3 +1,6 @@
+</span>
+{{#if is_bot}}
+    <i class="zulip-icon zulip-icon-bot" aria-label="{{t 'Bot' }}"></i>{{/if}}
 {{~!-- --~}}
```
Given that we have Handlebars whitespace control here, does this perturb the whitespace in a problematic way?

timabbott commented on web/styles/zulip.css:
```diff
@@ -2093,3 +2130,11 @@ body:not(.hide-left-sidebar) {
+
+.mention-content-wrapper .zulip-icon-bot {
+    margin-top: 100px;
+}
```
Where does this 100px magic number come from?

NotVishesh replied:
Sorry about this one, I was just testing few things and by mistake i committed this code .  I am really sorry

NotVishesh commented:
@timabbott  Thanks for wanting out those mistakes, I updated the PR please review

NotVishesh commented:
@timabbott  Thanks for wanting out those mistakes, I updated the PR please review

alya commented:
@amanagr would you be up for doing the next round of review on this one? I haven't tested.

amanagr commented:
@NotVishesh can you match the size of the bot icon with the adjacent icons like group icon in the left sidebar and in other places.

amanagr commented on web/src/pm_list_data.ts:
```diff
@@ -94,8 +94,9 @@ export function get_conversations(search_string = ""): DisplayObject[] {
-                // We display the bot icon rather than a user circle for bots.
                 is_bot = true;
+                // Green user circle is shown for bots.
+                user_circle_class = "user_circle_green";
```
grey circle?

amanagr reviewed (commented):
@NotVishesh I left some comments but I think the right way to make tangible progress here to just make this change in one part of the UI and we should try to merge that. Can you make a separate PR for just the message recipient bar?

NotVishesh commented:
@amanagr I Updated the PR with remaining four sections of the UI and updated the screenshots also.

amanagr commented on web/styles/user_circles.css:
```diff
@@ -41,3 +41,9 @@
     }
 }
+
+.user_circle_gray {
+    background-color: var(--grey-550);
+    border-radius: 50%;
+    border: 1px solid var(--grey-550);
+}
```
This needs a #design discussion, can you start a discussion for it in `#design`?

NotVishesh commented:
@amanagr  I updated the code and make the user circle green instead of gray for bots in left sidebar as told by Karl Stolley
in [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/Gray.20color.20indicator.20for.20bots)

amanagr commented on web/templates/pm_list_item.hbs:
```diff
@@ -3,15 +3,16 @@
             <span class="conversation-partners-list">{{recipients}}</span>
             {{> status_emoji status_emoji_info}}
+            {{#if is_bot}}
+                <i class="zulip-icon zulip-icon-bot conversation_bot_icon" aria-label="{{t 'Bot' }}"></i>
```
Do we need this `conversation_bot_icon` class here?

amanagr commented on web/src/user_pill.ts:
```diff
@@ -29,6 +29,7 @@ export type UserPill = {
     deactivated?: boolean;
     status_emoji_info?: (EmojiRenderingDetails & {emoji_alt_code?: boolean}) | undefined; // TODO: Move this in user_status.js
     should_add_guest_user_indicator?: boolean;
+    user_is_bot?: boolean;
```
Let's just name this `is_bot`.

amanagr commented:
Made a couple of comments, LGTM otherwise.

NotVishesh commented:
@amanagr Updated the PR

alya commented:
Please change it to use the offline dot, as specified in the issue, unless you have a specific reason why the green dot would work better. If you do, please raise it for discussion in the CZO thread.

NotVishesh commented:
I have updated the code for ```.user_circle_empty```

NotVishesh commented:
@amanagr please review.

amanagr commented:
Thanks, looks good to me now!

timabbott commented:
This spacing looks busted:

![image](https://github.com/user-attachments/assets/ceaf270d-8e00-4c8e-9c88-687182c2aaa5)

@NotVishesh did you re-test all of the places after the last revisions?

Also, can you make each commit message say it fixes part of #28561, and only the last one fix it? That way, it'll be possible to merge some if the others aren't finished.

NotVishesh commented:
@timabbott  Updated the PR with each commit saying fixes part of #28561.  And added a new commit for mention pill which saying it Fixes #26831 in which i adjusted the spacing of bot mention pill , which is now looking like 
![image](https://github.com/user-attachments/assets/3ca837df-1c45-4d1a-89d6-6b9e4268b164).

Also rechecked all my changes and that LGTM.

NotVishesh commented:
@timabbott please review

karlstolley commented:
@NotVishesh you should update your screenshots at the top of the PR demonstrating that there are no busted layouts anywhere in the UI. That will help get your PR reviewed much more quickly.

NotVishesh commented:
@karlstolley  Updated the all the Screenshots .

NotVishesh commented:
@timabbott please review.

alya commented:
Can you explain how you chose the colors for the icon? The sidebar UI one in particular looks a bit surprising to me.

NotVishesh commented:
@alya here is how it looks in both of the senarios 
<details>
<summary>bot icon with  previous used color in left Sidebar </summary>

![image](https://github.com/user-attachments/assets/c88a2829-2498-4ecc-a880-1cd2e698baea)

![image](https://github.com/user-attachments/assets/042eba1e-b06e-4055-95d0-a8c914163b19)


</details>

<details>
<summary>bot icon with  same color used elsewherer  </summary>

![image](https://github.com/user-attachments/assets/c9798352-a6ca-45b5-80af-60047c7daf22)

![image](https://github.com/user-attachments/assets/d11edd9a-e5f2-4949-97eb-94e2d7e2a127)


</details>

alya commented:
Could you please post both versions in #design for feedback?

NotVishesh commented:
@alya . Created a [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/color.20of.20bot.20icon.20in.20the.20left.20sidebar) for this

alya commented:
@timabbott adding this to release priorities because the way that bot pills look right now isn't really acceptable:
![Screenshot 2025-01-08 at 10 38 57](https://github.com/user-attachments/assets/433099d4-7283-4174-b5d1-6b8022a913f1)

timabbott commented on web/templates/recipient_row.hbs:
```diff
@@ -100,6 +100,9 @@
         </span>
+        {{#each recipient_users}}
+            {{#if is_bot}}<i class="zulip-icon zulip-icon-bot" aria-label="{{t 'Bot' }}"></i>{{/if}}
+        {{/each}}
```
I don't see how this being a loop can possibly make sense -- this needs to be inside the loop above, or you'll just get 3 bot icons in a row.

timabbott replied:
Yep, it's bugged.

![image](https://github.com/user-attachments/assets/ca3aa8d4-e393-44b9-96d5-8e677ac5a97d)

timabbott replied:
I dropped this commit, since I think it requires some added work to finish. @NotVishesh can you put up an updated PR with just that commit? I'm going to merge this one.

timabbott commented:
Merged, with the commit in https://github.com/zulip/zulip/pull/31894#discussion_r1909372356 dropped. Thanks @NotVishesh!

timabbott commented:
I opened #32990 to track that remaining work.
